### ISSUE TITLE
Missing HDF5 in Python 3.7 (not getting wheel on ReadTheDocs.org)

### ISSUE THREAD

**Issue by dwhswenson** [Unidata/netcdf4-python#1064]
I'm getting an error on installation of 1.5.5.1 when building my docs on RTD. It's trying to build from source, and not finding the HDF5 headers.

1.5.5 built fine. It looks like the difference is that wheels for CPython on x86_64 changed from `manylinux1` to `manylinux2014` between 1.5.5 and 1.5.5.1 (looking at the files on PyPI: [1.5.5](https://pypi.org/project/netCDF4/1.5.5/#files), [1.5.5.1](https://pypi.org/project/netCDF4/1.5.5.1/#files).) The last passing build downloaded the wheel. The failing build downloads the tgz and tries to build from there.

* Failing build: https://readthedocs.org/projects/openpathsampling-cli/builds/12686929/
* Last passing build: https://readthedocs.org/projects/openpathsampling-cli/builds/12584851/

There seem to be two problems:

1. If pip tries to install from source instead of from the wheel, it fails.
2. RTD isn't downloading the relevant wheel.

The first problem may just be expected, if the relevant files aren't expected to be included in the tgz. I think second problem would be solved if netcdf4 added a `manylinux1` wheel.

Comparing a couple other projects (for CPython and x86_64, behavior differs for aarch64 or pypy):

* [numpy 1.19.4 builds `manylinux1` and `manylinux2010`](https://pypi.org/project/numpy/#files)
* [matplotlib 3.3.3 only builds `manylinux1`](https://pypi.org/project/matplotlib/#files)
* [pandas 1.2.0 only builds `manylinux1`](https://pypi.org/project/pandas/#files) 

<details> <summary>Pip output and traceback of failing build</summary>

```
Searching for netcdf4
Reading https://pypi.org/simple/netcdf4/
Downloading https://files.pythonhosted.org/packages/9f/e2/686d0d937f979e7ee0eb6bb4821448da6d2910789f4928c12b070f424b6a/netCDF4-1.5.5.1.tar.gz#sha256=d957e55a667d1fc651ddef22fea10ded0f142b7d9dbbf4d08c0012d32f445abd
Best match: netCDF4 1.5.5.1
Processing netCDF4-1.5.5.1.tar.gz
Writing /tmp/easy_install-50y5flv4/netCDF4-1.5.5.1/setup.cfg
Running netCDF4-1.5.5.1/setup.py -q bdist_egg --dist-dir /tmp/easy_install-50y5flv4/netCDF4-1.5.5.1/egg-dist-tmp-j4aiy83n
Package hdf5 was not found in the pkg-config search path.
Perhaps you should add the directory containing `hdf5.pc'
to the PKG_CONFIG_PATH environment variable
No package 'hdf5' found
reading from setup.cfg...

    HDF5_DIR environment variable not set, checking some standard locations ..
checking /home/docs/include ...
hdf5 headers not found in /home/docs/include
checking /usr/local/include ...
hdf5 headers not found in /usr/local/include
checking /sw/include ...
hdf5 headers not found in /sw/include
checking /opt/include ...
hdf5 headers not found in /opt/include
checking /opt/local/include ...
hdf5 headers not found in /opt/local/include
checking /usr/include ...
hdf5 headers not found in /usr/include
```
```pytb
Traceback (most recent call last):
  File "/home/docs/checkouts/readthedocs.org/user_builds/openpathsampling-cli/envs/latest/lib/python3.7/site-packages/setuptools/sandbox.py", line 154, in save_modules
    yield saved
  File "/home/docs/checkouts/readthedocs.org/user_builds/openpathsampling-cli/envs/latest/lib/python3.7/site-packages/setuptools/sandbox.py", line 195, in setup_context
    yield
  File "/home/docs/checkouts/readthedocs.org/user_builds/openpathsampling-cli/envs/latest/lib/python3.7/site-packages/setuptools/sandbox.py", line 250, in run_setup
    _execfile(setup_script, ns)
  File "/home/docs/checkouts/readthedocs.org/user_builds/openpathsampling-cli/envs/latest/lib/python3.7/site-packages/setuptools/sandbox.py", line 45, in _execfile
    exec(code, globals, locals)
  File "/tmp/easy_install-50y5flv4/netCDF4-1.5.5.1/setup.py", line 411, in <module>
  File "/tmp/easy_install-50y5flv4/netCDF4-1.5.5.1/setup.py", line 356, in _populate_hdf5_info
ValueError: did not find HDF5 headers

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/docs/checkouts/readthedocs.org/user_builds/openpathsampling-cli/checkouts/latest/setup.py", line 128, in <module>
    setup()
  File "/home/docs/checkouts/readthedocs.org/user_builds/openpathsampling-cli/envs/latest/lib/python3.7/site-packages/setuptools/__init__.py", line 145, in setup
    return distutils.core.setup(**attrs)
  File "/home/docs/.pyenv/versions/3.7.9/lib/python3.7/distutils/core.py", line 148, in setup
    dist.run_commands()
  File "/home/docs/.pyenv/versions/3.7.9/lib/python3.7/distutils/dist.py", line 966, in run_commands
    self.run_command(cmd)
  File "/home/docs/.pyenv/versions/3.7.9/lib/python3.7/distutils/dist.py", line 985, in run_command
    cmd_obj.run()
  File "/home/docs/checkouts/readthedocs.org/user_builds/openpathsampling-cli/envs/latest/lib/python3.7/site-packages/setuptools/command/install.py", line 67, in run
    self.do_egg_install()
  File "/home/docs/checkouts/readthedocs.org/user_builds/openpathsampling-cli/envs/latest/lib/python3.7/site-packages/setuptools/command/install.py", line 117, in do_egg_install
    cmd.run()
  File "/home/docs/checkouts/readthedocs.org/user_builds/openpathsampling-cli/envs/latest/lib/python3.7/site-packages/setuptools/command/easy_install.py", line 418, in run
    self.easy_install(spec, not self.no_deps)
  File "/home/docs/checkouts/readthedocs.org/user_builds/openpathsampling-cli/envs/latest/lib/python3.7/site-packages/setuptools/command/easy_install.py", line 660, in easy_install
    return self.install_item(None, spec, tmpdir, deps, True)
  File "/home/docs/checkouts/readthedocs.org/user_builds/openpathsampling-cli/envs/latest/lib/python3.7/site-packages/setuptools/command/easy_install.py", line 707, in install_item
    self.process_distribution(spec, dist, deps)
  File "/home/docs/checkouts/readthedocs.org/user_builds/openpathsampling-cli/envs/latest/lib/python3.7/site-packages/setuptools/command/easy_install.py", line 752, in process_distribution
    [requirement], self.local_index, self.easy_install
  File "/home/docs/checkouts/readthedocs.org/user_builds/openpathsampling-cli/envs/latest/lib/python3.7/site-packages/pkg_resources/__init__.py", line 782, in resolve
    replace_conflicting=replace_conflicting
  File "/home/docs/checkouts/readthedocs.org/user_builds/openpathsampling-cli/envs/latest/lib/python3.7/site-packages/pkg_resources/__init__.py", line 1065, in best_match
    return self.obtain(req, installer)
  File "/home/docs/checkouts/readthedocs.org/user_builds/openpathsampling-cli/envs/latest/lib/python3.7/site-packages/pkg_resources/__init__.py", line 1077, in obtain
    return installer(requirement)
  File "/home/docs/checkouts/readthedocs.org/user_builds/openpathsampling-cli/envs/latest/lib/python3.7/site-packages/setuptools/command/easy_install.py", line 679, in easy_install
    return self.install_item(spec, dist.location, tmpdir, deps)
  File "/home/docs/checkouts/readthedocs.org/user_builds/openpathsampling-cli/envs/latest/lib/python3.7/site-packages/setuptools/command/easy_install.py", line 705, in install_item
    dists = self.install_eggs(spec, download, tmpdir)
  File "/home/docs/checkouts/readthedocs.org/user_builds/openpathsampling-cli/envs/latest/lib/python3.7/site-packages/setuptools/command/easy_install.py", line 890, in install_eggs
    return self.build_and_install(setup_script, setup_base)
  File "/home/docs/checkouts/readthedocs.org/user_builds/openpathsampling-cli/envs/latest/lib/python3.7/site-packages/setuptools/command/easy_install.py", line 1158, in build_and_install
    self.run_setup(setup_script, setup_base, args)
  File "/home/docs/checkouts/readthedocs.org/user_builds/openpathsampling-cli/envs/latest/lib/python3.7/site-packages/setuptools/command/easy_install.py", line 1144, in run_setup
    run_setup(setup_script, args)
  File "/home/docs/checkouts/readthedocs.org/user_builds/openpathsampling-cli/envs/latest/lib/python3.7/site-packages/setuptools/sandbox.py", line 253, in run_setup
    raise
  File "/home/docs/.pyenv/versions/3.7.9/lib/python3.7/contextlib.py", line 130, in __exit__
    self.gen.throw(type, value, traceback)
  File "/home/docs/checkouts/readthedocs.org/user_builds/openpathsampling-cli/envs/latest/lib/python3.7/site-packages/setuptools/sandbox.py", line 195, in setup_context
    yield
  File "/home/docs/.pyenv/versions/3.7.9/lib/python3.7/contextlib.py", line 130, in __exit__
    self.gen.throw(type, value, traceback)
  File "/home/docs/checkouts/readthedocs.org/user_builds/openpathsampling-cli/envs/latest/lib/python3.7/site-packages/setuptools/sandbox.py", line 166, in save_modules
    saved_exc.resume()
  File "/home/docs/checkouts/readthedocs.org/user_builds/openpathsampling-cli/envs/latest/lib/python3.7/site-packages/setuptools/sandbox.py", line 141, in resume
    six.reraise(type, exc, self._tb)
  File "/home/docs/checkouts/readthedocs.org/user_builds/openpathsampling-cli/envs/latest/lib/python3.7/site-packages/setuptools/_vendor/six.py", line 685, in reraise
    raise value.with_traceback(tb)
  File "/home/docs/checkouts/readthedocs.org/user_builds/openpathsampling-cli/envs/latest/lib/python3.7/site-packages/setuptools/sandbox.py", line 154, in save_modules
    yield saved
  File "/home/docs/checkouts/readthedocs.org/user_builds/openpathsampling-cli/envs/latest/lib/python3.7/site-packages/setuptools/sandbox.py", line 195, in setup_context
    yield
  File "/home/docs/checkouts/readthedocs.org/user_builds/openpathsampling-cli/envs/latest/lib/python3.7/site-packages/setuptools/sandbox.py", line 250, in run_setup
    _execfile(setup_script, ns)
  File "/home/docs/checkouts/readthedocs.org/user_builds/openpathsampling-cli/envs/latest/lib/python3.7/site-packages/setuptools/sandbox.py", line 45, in _execfile
    exec(code, globals, locals)
  File "/tmp/easy_install-50y5flv4/netCDF4-1.5.5.1/setup.py", line 411, in <module>
  File "/tmp/easy_install-50y5flv4/netCDF4-1.5.5.1/setup.py", line 356, in _populate_hdf5_info
ValueError: did not find HDF5 headers
```

</details>

**Comment by jswhit** [comment]
I don't see why it would not install the manylinux2014 wheel unless you are running an old version of pip (pip < 19.3).  Can you try again after upgrading pip?

I can provide manylinux1 wheels, but the problem is that the latest version of the supporting C libs won't build on that old a version, so some features may be missing.

**Comment by dwhswenson** [comment]
> I don't see why it would not install the manylinux2014 wheel unless you are running an old version of pip (pip < 19.3). Can you try again after upgrading pip?

This is running in containers on ReadTheDocs.org. I can add artificial requirements to my code, but I can't change their default behavior. In any case, according to RTD, they're running pip 20.3.3 (the current version). From the logs of the failing build:

```
/home/docs/checkouts/readthedocs.org/user_builds/openpathsampling-cli/envs/latest/bin/python -m pip install --upgrade --no-cache-dir pip
```

```
Requirement already satisfied: pip in /home/docs/checkouts/readthedocs.org/user_builds/openpathsampling-cli/envs/latest/lib/python3.7/site-packages (20.3.3)
```

Maybe they're using CentOS 6 containers? I see them pulling in wheels with `manylinux2010` or `manylinux1`, but none with `manylinux2014`.

**Comment by jswhit** [comment]
Hmm.  Maybe CentOS 6 requires manylinux2010 wheels.  I will try building those and uploading to pypi.

**Comment by jswhit** [comment]
Uploaded manylinux2010 wheels for python 3.6,3.7,3.8 and 3.9 and manylinux1 wheels for 3.6.  Let me know if that fixes the issue on RTD.

**Comment by jswhit** [comment]
Can this issue be closed?  @dwhswenson - have you tried this since the manylinux2010 wheels were uploaded to PyPI?

**Comment by dwhswenson** [comment]
It didn't fix, but that's because I had misunderstood some things. I've now managed to fix it on my end, but it's worth explaining the problem here in case others bump against this. This is definitely an issue that came up because netCDF4 is on the leading edge in terms of dropping old wheel formats. RTD seems to pin things to older versions, and the problem occurs when these two policies clash. Details:

**Corrections:** I had misspoken before about the existence of `manylinux2010` builds in this process: They do get downloaded by some stages of the build process, but importantly, *not the stage that is failing*. Another thing I had been wrong about: the failing stage is a `python setup.py install`, not a `pip install`.

**Fundamental problem:** For some reason, RTD pins `setuptools=41.0.1` (and `sphinx<2`?!? I'm pretty sure I didn't ask for that). Support for `manylinux2010` wheels was added to setuptools in 42.0.0. I'm honestly quite confused about how RTD selects versions by default; I seem to remember seeing a relatively recent project (perhaps the same one) was trying to build docs on Python 2.7.

**Solution:** Adding a minimal [`.readthedocs.yml`](https://docs.readthedocs.io/en/stable/config-file/v2.html) config file which installs via `pip` (pretty much identical to the example at that link) fixed it. Since the `pip` version is up-to-date, it should take any supported wheels.

I guess this is just a difference in policy between netCDF4 and RTD. It's easily fixed on the user end (by doing what RTD recommends anyway: adding a config file for RTD), so I'll close this. Thanks for the help @jswhit!
